File: assignment/assignment1/scala/hw1/src/main/scala/task2.scala
/*
Scala: $ spark-submit --class task2 hw1.jar <review_file> <business_file> <output_file> <if_spark> <n>

$ sbt package
// spark
$ spark-submit --class task2 ./target/scala-2.11/hw1_2.11-0.1-SNAPSHOT.jar "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/business.json" "output_task2_spark_scala.json" "spark" 30
// no_spark
$ spark-submit --class task2 ./target/scala-2.11/hw1_2.11-0.1-SNAPSHOT.jar "/Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "/Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/business.json" "output_task2_nospark_scala.json" "no_spark" 30
*/

import scala.io.Source
import java.io.PrintWriter
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf

import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization



object task2 {
    def main(args: Array[String]) {
        // parameters
        val review_file = args(0)
        val business_file =args(1)
        val output_file = args(2)
        val if_spark = args(3)
        val n = args(4).toInt

        def withSpark() {
            val conf = new SparkConf().setAppName("task2").setMaster("local[*]")
            val sc = new SparkContext(conf)

            def textToMap(r: String): Map[String, Any] = {
                implicit val formats = DefaultFormats
                parse(r).extract[Map[String, Any]]
            }

            def businessWithOneCategory(t: (String, String)): List[(String, String)] = {
                val business: String = t._1
                val category: String = t._2
                val cate_array: Array[String] = category.split(",").map(_.trim())
                var bc = List[(String, String)]()
                for (c <- cate_array) {
                    bc = bc:::List((business, c))
                }
                bc
            }
            val business_data = sc.textFile(business_file)
                                .map(textToMap)
                                .filter(b => b("categories") != null)
                                .flatMap(b => businessWithOneCategory((b("business_id").toString, b("categories").toString)))
            // business_data.foreach(println)

            val review_data = sc.textFile(review_file)
                                .map(textToMap)
                                .map(r => (r("business_id").toString, (r("stars").asInstanceOf[Number].floatValue, 1)))
                                .reduceByKey((x, y) => ((x._1+y._1), (x._2+y._2)))
            // review_data.foreach(println)

            val data_join_on_bussiness_id = business_data.join(review_data)
                                                        .map(_._2)
                                                        .reduceByKey((x, y) => ((x._1+y._1), (x._2+y._2)))
                                                        .mapValues(t => t._1/t._2)
                                                        .sortBy(_._1, ascending=true)
                                                        .sortBy(_._2, ascending=false)
                                                        .map(x => List(x._1, x._2))
                                                        .take(n)
            data_join_on_bussiness_id.foreach(println)

            // result
            val result = Map("result" -> data_join_on_bussiness_id)
            implicit val formats = DefaultFormats
            val result_json = Serialization.write(result)
            println(result_json)
            
            // write to output_task2_scala.json file
            // val output = "output_task2_scala.json"
            val out = new PrintWriter(output_file)
            out.print(result_json)
            out.close()
        }

        def withoutSpark() {
            // process review.json
            val review_file_pointer = Source.fromFile(review_file)
            val review_file_lines = review_file_pointer.getLines

            def processReviewLines(r: String): (String, (Float, Int)) = {
                implicit val formats = DefaultFormats
                val rm = parse(r).extract[Map[String, Any]]
                (rm("business_id").toString, (rm("stars").asInstanceOf[Number].floatValue, 1))
            }

            def sumAndCountValue(x:(String, List[(String, (Float, Int))])): (String, (Float, Int))= {
                val key = x._1
                val l = x._2
                var sum: Float = 0.0f
                var count: Int = 0
                for (item <- l) {
                    sum += item._2._1
                    count += item._2._2
                }
                (key, (sum, count))
            }
            val review_data_map = review_file_lines.map(processReviewLines)
                                            .toList
                                            .groupBy(_._1)
                                            .map(x => sumAndCountValue(x))
                                            .toMap
            // println(review_data_map)
            
            // process business.json
            val business_file_pointer = Source.fromFile(business_file)
            val business_file_lines = business_file_pointer.getLines

            def parseBusinessLines(b: String): Map[String, Any] = {
                implicit val formats = DefaultFormats
                parse(b).extract[Map[String, Any]]
            }

            def processBusinessData(b: Map[String, Any]): List[(String, String)] = {
                val business = b("business_id").toString
                val categories = b("categories").toString
                val cate_array = categories.split(",").map(_.trim())
                var bc = List[(String, String)]()
                for (c <- cate_array) {
                    bc = bc:::List((c, business))
                }
                bc
            }
            val business_data = business_file_lines.map(parseBusinessLines)
                                                .filter(b => b("categories") != null)
                                                .flatMap(processBusinessData)
                                                .filter(x => review_data_map.get(x._2) != None)
            
            // join business and review on business_id
            def getSumCountFromReview(b: (String, String)): (String, (Float, Int)) = {
                val category = b._1
                val business = b._2
                val sum_count = review_data_map(business)
                (category, sum_count)
            }

            val result_after_join = business_data.map(getSumCountFromReview)
                                                .toList
                                                .groupBy(_._1)
                                                .map(sumAndCountValue)
                                                .map(x => (x._1, x._2._1/x._2._2))
                                                .toList
                                                .sortBy(_._1).reverse
                                                .sortBy(_._2).reverse
                                                .map(x => List(x._1, x._2))
                                                .take(n)

            // result
            val result = Map("result" -> result_after_join)
            implicit val formats = DefaultFormats
            val result_json = Serialization.write(result)
            println(result_json)
            
            // write to output file
            val out = new PrintWriter(output_file)
            out.print(result_json)
            out.close()
        }
        
        if (if_spark == "spark"){
            withSpark()
        } else if (if_spark == "no_spark") {
            withoutSpark()
        } else {
            println("Wrong if_spark value.")
        }
    }
}

File: assignment/assignment1/python/task2.py
"""
Python: $ spark-submit task2.py <review_file> <business_file > <output_file> <if_spark> <n>
spark:
spark-submit task2.py "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/business.json" "output_task2.json" "spark" 5
no_spark:
spark-submit task2.py "/Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "/Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/business.json" "output_task2.json" "no_spark" 5

review _file – the input file (the review dataset) 
business_file – the input file (the business dataset) 
output_file – the output file contains your answers 
if_spark – either “spark” or “no_spark” 
n – top n categories with highest average stars (see 4.2.1)
"""
import json
import time
import sys

review_file = sys.argv[1]
business_file = sys.argv[2]
output_file = sys.argv[3]
if_spark = sys.argv[4]
n = int(sys.argv[5])

def timer(func):
    """timer decorator"""
    def wrapper(*arg, **karg):
        t1 = time.time()
        res = func(*arg, **karg)
        t2 = time.time()
        delta_t = t2 - t1
        print('time consume:%.2fs' % delta_t)
        return res
    return wrapper

def write_output_file(result):
    # write on local disk instead of hdfs
    with open(output_file, 'w', encoding='utf-8') as fp:
        json.dump(result, fp)

@timer
def noSpark():
    from functools import reduce

    def openDataFile(file_addr, a, b):
        with open(file_addr, 'r', encoding='utf-8') as fp:
            data = []
            for line in fp.readlines():
                d_= json.loads(line)
                data.append({a: d_[a], b: d_[b]})
        return(data)

    review_data = openDataFile(review_file, 'business_id', 'stars')
    sub_review_data = [[x['business_id'], [x['stars']]] for x in review_data] # [[business_id, [starts]]...]
    sub_review_data.sort(key=lambda x: x[0])
    def groupBy(r_l):
        li = [r_l[0]]
        for i in range(1, len(r_l)):
            key_ = r_l[i][0]
            value_ = r_l[i][1]
            if key_ == r_l[i-1][0]:
                li[-1][1] += value_
            else:
                li.append(r_l[i])
        return li
    grouped_sub_review_data = groupBy(sub_review_data)
    sub_review_data_combine_list = [[x[0], [sum(x[1]), len(x[1])]] for x in grouped_sub_review_data]
    sub_review_data_combine_dict = {}
    for x in sub_review_data_combine_list:
        sub_review_data_combine_dict[x[0]] = x[1]

    business_data = openDataFile(business_file, 'business_id', 'categories')
    filtered_business_data = [x for x in business_data if x['categories'] != None]
    sub_business_data = [[x['business_id'], [x.strip() for x in x['categories'].split(',')]] for x in filtered_business_data]
    sub_business_data_flat = [[x, item[0]] for item in sub_business_data for x in item[1]] # [[category, business_id], ...]
    sub_business_data_flat.sort(key=lambda x: x[0])
 
    def joinBR(bl, rd):
        """
        bl [[category, business_id], ...]
        rd {business_id: [sum, count], ...}
        """
        l_ = []
        for i in range(len(bl)):
            key_ = bl[i][0]
            value_ = rd.get(bl[i][1])
            if value_ == None:
                continue
            else:
                l_.append([key_, value_])
        return l_
    
    joined_b_r = joinBR(sub_business_data_flat, sub_review_data_combine_dict) # [[category, [sum_stars, 5]], ...]
    def groupByJoinedBR(jl):
        """
        jl: [[category, [sum_stars, 5]], ...]
        """
        jl1 = [[x[0], [x[1]]] for x in jl]
        jl1.sort(key=lambda x: x[0])
        li = [jl1[0]]
        for i in range(1, len(jl1)):
            k_ = jl1[i][0]
            v_ = jl1[i][1]
            if k_ == jl1[i-1][0]:
                li[-1][1] += v_
            else:
                li.append(jl1[i])
        return li
    groupedbr = groupByJoinedBR(joined_b_r)

    def reducefunc(scl):
        x_0 = scl[0][0]
        x_1 = scl[0][1]
        for j in range(1, len(scl)):
            v_0 = scl[j][0]
            v_1 = scl[j][1]
            
            x_0 += v_0
            x_1 += v_1
        result = x_0/x_1
        return result

    for x in groupedbr:
        x[1] = reducefunc(x[1])

    result_sort = groupedbr
    result_sort.sort(key=lambda x: x[1], reverse=True)
    result = {"result": result_sort[:n]}

    write_output_file(result)

@timer
def spark():
    from pyspark import SparkContext

    # define spark env
    sc = SparkContext(master="local[*]", appName="task2")

    business_data = sc.textFile(business_file).cache()
    sub_business_data = business_data.map(json.loads) \
        .filter(lambda b: b['categories'] != None) \
        .map(lambda b: (b['business_id'], [x.strip() for x in b['categories'].split(',')])) \
        .flatMap(lambda item: [(item[0], x) for x in item[1]])
    # print(sub_business_data)

    review_data = sc.textFile(review_file).cache()
    sub_review_data = review_data.map(json.loads) \
        .map(lambda r: (r['business_id'], (r['stars'], 1))) \
        .reduceByKey(lambda x, y: (x[0]+y[0], x[1]+y[1]))
    # print(sub_review_data)

    data_join_on_bussiness_id = sub_business_data.join(sub_review_data) \
        .map(lambda x: x[1]) \
        .reduceByKey(lambda x, y: [x[0]+y[0], x[1]+y[1]]) \
        .mapValues(lambda l: l[0]/l[1]) \
        .sortBy(keyfunc=lambda x: x[0], ascending=True) \
        .sortBy(keyfunc=lambda x: x[1], ascending=False) \
        .map(lambda x: [x[0], x[1]]) \
        .take(n)
    # print(data_join_on_bussiness_id)

    result = {'result': data_join_on_bussiness_id}
    write_output_file(result)

if __name__ == "__main__":
    if if_spark == "spark":
        spark()
    elif if_spark == "no_spark":
        noSpark()
    else:
        print('wrong if_spark value.')

